internal/validation/apiextensions/v1/composition/handler.go

```diff
@@ -74,103 +67,91 @@ func SetupWebhookWithManager(mgr ctrl.Manager, options controller.Options) error
 		}
 	}
 
-	// TODO(lsviben): switch to using admission.CustomValidator when https://github.com/kubernetes-sigs/controller-runtime/issues/1896 is resolved.
-	mgr.GetWebhookServer().Register(v1.CompositionValidatingWebhookPath,
-		&webhook.Admission{Handler: &handler{
-			reader:  unstructured.NewClient(mgr.GetClient()),
```

Comment:
I removed the `unstructured.NewClient` that wraps the manager's client here. I can't see any reason we'd need it. Let me know if I missed it.

Reply:
I think it was needed for rendered resource validation IIRC, fine removing it now, I'll have to manually test it though, would be nice to have an e2e for this.

Reply:
@phisco I added a rudimentary E2E test, which seems to pass. Can you take a look and confirm whether it's actually testing anything useful? Maybe I should enable strict validation on the relevant `Composition`.

Reply:
The test case looks good to me as an initial smoke test, I would maybe just add another step checking that an invalid composition is rejected, e.g.:
Change the XRD to have one more `evenCoolerField` of type integer:
```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xnopresources.nop.example.org
spec:
  group: nop.example.org
  names:
    kind: XNopResource
    plural: xnopresources
  claimNames:
    kind: NopResource
    plural: nopresources
  connectionSecretKeys:
  - test
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
     openAPIV3Schema:
       type: object
       properties:
        spec:
          type: object
          properties:
            coolField:
              type: string
          required:
          - coolField
        status:
          type: object
          properties:
            coolerField:
              type: string
            evenCoolerField:               # <== added field
              type: integer
```
which then we are going to patch from a string without any transformation:
```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: xnopresources.nop.example.org
spec:
  compositeTypeRef:
    apiVersion: nop.example.org/v1alpha1
    kind: XNopResource
  resources:
  - name: nop-resource-1
    base:
     apiVersion: nop.crossplane.io/v1alpha1
     kind: NopResource
     spec:
      forProvider:
        conditionAfter:
        - conditionType: Ready
          conditionStatus: "False"
          time: 0s
        - conditionType: Ready
          conditionStatus: "True"
          time: 10s
    patches:
    - type: FromCompositeFieldPath
      fromFieldPath: spec.coolField
      toFieldPath: metadata.annotations[cool-field]
      transforms:
      - type: string
        string:
         type: Convert
         convert: ToUpper
    - type: ToCompositeFieldPath
      fromFieldPath: metadata.annotations[cool-field]
      toFieldPath: status.coolerField
    - type: ToCompositeFieldPath                                # <== added patch
      fromFieldPath: metadata.annotations[cool-field]
      toFieldPath: status.evenCoolerField
```
And once applied we should get an error, I wouldn't check here the specific error, as that's checked by the unit tests, but manually applying it results in the following error:
```
The Composition "xnopresources.nop.example.org" is invalid: spec.resources[0].patches[2].transforms: Required value: the fromFieldPath does not have a type compatible with the toFieldPath according to their schemas and no transforms were provided: string != integer
```

Reply:
Done!